api/query/cache/monitor/monitor.go

```diff
@@ -114,19 +116,36 @@ func (m *indexMonitor) Start() error {
 		return err
 	}
 
+	m.idx.SetIngestChan(m.runIngested)
+
+	reqs := uint(0)
 	for {
 		if !m.isRunning {
 			return errStopped
 		}
 
-		heapBytes := m.rt.GetHeapBytes()
-		if heapBytes > m.maxHeapBytes {
-			m.logger.Errorf("Out of memory: %d > %d", heapBytes, m.maxHeapBytes)
-			m.idx.EvictRuns(m.percent)
-		} else {
-			m.logger.Debugf("Monitor: %d heap-allocated bytes OK", heapBytes)
+		timer := make(chan bool, 1)
+		go func() {
+			time.Sleep(m.interval)
+			timer <- true
+		}()
```

Comment:
I think you'd need to extract this channel and goroutine out of the infinite loop and add an infinite loop within the function; i.e., have a single, long-living goroutine that writes to the same `timer` channel periodically.

Imagine if you have runs coming in such a fast rate that eviction is never triggered by the timer, you'd end up with a whole bunch of unnecessary goroutines sleeping. Besides, `timer` is created in each iteration but never closed, and I'm not sure if this would confuse the GC.

Reply:
I don't think there's a GC issue. There are no references to `timer` after this goroutine exits.

The trouble with a separate goroutine is that it doesn't start its timeout after the last `check()`, it starts it independently. That means the monitor may run more frequently than intended.